### ISSUE TITLE
AOT compile dependencies

### ISSUE THREAD

**Issue by anolivetree** [gradle-clojure/gradle-clojure#52]
Will you add a feature to AOT compile clj files in jar files? Just like Leiningen's :aot :all.

**Comment by pbzdyl** [comment]
Hello @anolivetree,

Have you tried the [following](https://github.com/gradle-clojure/gradle-clojure/blob/master/docs/README.md#clojurecompile):

```
compileClojure {
  options.aotCompile = true
}
```

in `build.gradle`?

When you set `options.aotCompile` without providing `options.namespaces` explicitly, then all namespaces will be AOT compiled. If you would like to AOT only selected namespaces, then specify them explicitly:

```
compileClojure {
  options {
    aotCompile = true
    namespaces = ["myapp.core", "myapp.utils"]
  }
}
```

**Comment by anolivetree** [comment]
I tried again. It doesn't work in some cases.

After I killed all gradle processes and do,
$ ./gradlew clean
$ ./gradlew assemble
the created jar contains 3rd party classes too as expected.

However, if I do follwing again with the gradle process running launched on the first try,
$ ./gradlew clean
$ ./gradlew assemble
the created jar doesn't contain 3rd party classes.

**Comment by pbzdyl** [comment]
@anolivetree I tried to replicate your issue by:
- clone https://github.com/gradle-clojure/gradle-clojure-samples
- add clj-http and clj-time
- require them in the core ns
- add `compileClojure { options.aotCompile = true }` in build.gradle

Then I tried to run `./gradlew clean` and `./gradlew -x check assemble` repeatedly but I couldn't get the jar built without AOT compiled classes from clj-http and clj-time.

Could you please share a copy of a project that recreates the issue?

Which gradle version do you use?

**Comment by anolivetree** [comment]
Hi @pbzdyl 

I've just pushed the source code.
https://github.com/anolivetree/gradle-clojure-samples

Following is the exact command I used. 
On first build, created jar contains httpkit. However, 2nd build doesn't include httpkit.

```
sakurai@envyhome ~/project/olive/clojure/gradle-clojure-samples $ ps ax |grep gradle
27257 pts/5    T      0:00 vim build.gradle
29253 pts/2    S+     0:00 grep --color=auto gradle
sakurai@envyhome ~/project/olive/clojure/gradle-clojure-samples $ ./gradlew clean
Starting a Gradle Daemon, 1 incompatible and 5 stopped Daemons could not be reused, use --status for details

BUILD SUCCESSFUL in 2s
1 actionable task: 1 up-to-date
sakurai@envyhome ~/project/olive/clojure/gradle-clojure-samples $ ps ax |grep gradle
27257 pts/5    T      0:00 vim build.gradle
29287 ?        Ssl    0:08 /usr/lib/jvm/java-8-oracle/bin/java -XX:MaxPermSize=512m -Dfile.encoding=UTF-8 -Duser.country=US -Duser.language=en -Duser.variant -cp /home/sakurai/.gradle/wrapper/dists/gradle-4.3-bin/452wx51oxqsia28686mgqhot6/gradle-4.3/lib/gradle-launcher-4.3.jar org.gradle.launcher.daemon.bootstrap.GradleDaemon 4.3
29345 pts/2    S+     0:00 grep --color=auto gradle
sakurai@envyhome ~/project/olive/clojure/gradle-clojure-samples $ ./gradlew assemble

> Task :compileClojure 
Creating fresh Clojure runtime.


BUILD SUCCESSFUL in 1s
3 actionable tasks: 3 executed
sakurai@envyhome ~/project/olive/clojure/gradle-clojure-samples $ unzip -l build/libs/gradle-clojure-samples-0.1.0-SNAPSHOT.jar 
Archive:  build/libs/gradle-clojure-samples-0.1.0-SNAPSHOT.jar
  Length      Date    Time    Name
---------  ---------- -----   ----
        0  2017-11-28 22:33   META-INF/
       25  2017-11-28 22:33   META-INF/MANIFEST.MF
        0  2017-11-28 22:33   sample/
     2057  2017-11-28 22:33   sample/core$loading__5569__auto____32.class
     1142  2017-11-28 22:33   sample/core$bar.class
     2783  2017-11-28 22:33   sample/core__init.class
      974  2017-11-28 22:33   sample/core$foo.class
     1322  2017-11-28 22:33   sample/core$fn__196.class
        0  2017-11-28 22:33   org/
        0  2017-11-28 22:33   org/httpkit/
      952  2017-11-28 22:33   org/httpkit/client$query_string$join__61.class
     1712  2017-11-28 22:33   org/httpkit/client$nested_param$fn__50.class
      934  2017-11-28 22:33   org/httpkit/client$query_string$iter__63__67.class
     2351  2017-11-28 22:33   org/httpkit/client$head.class
     1349  2017-11-28 22:33   org/httpkit/client$fn__36.class
     2277  2017-11-28 22:33   org/httpkit/client$query_string$iter__63__67$fn__68$fn__69.class
     5762  2017-11-28 22:33   org/httpkit/client$defreq.class
     1337  2017-11-28 22:33   org/httpkit/client$request$deliver_resp__109.class
     1293  2017-11-28 22:33   org/httpkit/client$query_string.class
     7115  2017-11-28 22:33   org/httpkit/client$request.class
     2349  2017-11-28 22:33   org/httpkit/client$put.class
     1069  2017-11-28 22:33   org/httpkit/client$url_encode.class
     6199  2017-11-28 22:33   org/httpkit/client$request$reify__114.class
     1650  2017-11-28 22:33   org/httpkit/client$basic_auth_value.class
     1046  2017-11-28 22:33   org/httpkit/client$nested_param.class
     5247  2017-11-28 22:33   org/httpkit/client$coerce_req.class
     2349  2017-11-28 22:33   org/httpkit/client$get.class
     2351  2017-11-28 22:33   org/httpkit/client$copy.class
     2351  2017-11-28 22:33   org/httpkit/client$post.class
     3364  2017-11-28 22:33   org/httpkit/client$query_string$iter__63__67$fn__68.class
     1939  2017-11-28 22:33   org/httpkit/client$coerce_req$fn__85.class
     2361  2017-11-28 22:33   org/httpkit/client$proppatch.class
     1705  2017-11-28 22:33   org/httpkit/client$fn__98.class
     2359  2017-11-28 22:33   org/httpkit/client$propfind.class
     2357  2017-11-28 22:33   org/httpkit/client$options.class
     2355  2017-11-28 22:33   org/httpkit/client$delete.class
      950  2017-11-28 22:33   org/httpkit/client$coerce_req$fn__89.class
     2355  2017-11-28 22:33   org/httpkit/client$report.class
     1012  2017-11-28 22:33   org/httpkit/client$utf8_bytes.class
      761  2017-11-28 22:33   org/httpkit/client$max_body_filter.class
      573  2017-11-28 22:33   org/httpkit/client$fn__98$fn__99.class
     2349  2017-11-28 22:33   org/httpkit/client$acl.class
     2355  2017-11-28 22:33   org/httpkit/client$unlock.class
    16509  2017-11-28 22:33   org/httpkit/client__init.class
     1327  2017-11-28 22:33   org/httpkit/client$prepare_response_headers$fn__46.class
      729  2017-11-28 22:33   org/httpkit/client$base64_encode.class
     2353  2017-11-28 22:33   org/httpkit/client$patch.class
     2823  2017-11-28 22:33   org/httpkit/client$prepare_request_headers.class
     2058  2017-11-28 22:33   org/httpkit/client$request$deliver_resp__109$fn__110.class
     2351  2017-11-28 22:33   org/httpkit/client$move.class
     2351  2017-11-28 22:33   org/httpkit/client$lock.class
     3568  2017-11-28 22:33   org/httpkit/client$loading__5569__auto____34.class
     1251  2017-11-28 22:33   org/httpkit/client$query_string$param__59.class
     1181  2017-11-28 22:33   org/httpkit/client$prepare_response_headers.class
     1729  2017-11-28 22:33   org/httpkit/client$nested_param$fn__50$fn__53.class
     1317  2017-11-28 22:33   org/httpkit/client$fn__95.class
       11  2017-11-28 22:33   sample.txt
---------                     -------
   124349                     57 files
sakurai@envyhome ~/project/olive/clojure/gradle-clojure-samples $ ./gradlew clean

BUILD SUCCESSFUL in 0s
1 actionable task: 1 executed
sakurai@envyhome ~/project/olive/clojure/gradle-clojure-samples $ ps ax |grep gradle
27257 pts/5    T      0:00 vim build.gradle
29287 ?        Ssl    0:09 /usr/lib/jvm/java-8-oracle/bin/java -XX:MaxPermSize=512m -Dfile.encoding=UTF-8 -Duser.country=US -Duser.language=en -Duser.variant -cp /home/sakurai/.gradle/wrapper/dists/gradle-4.3-bin/452wx51oxqsia28686mgqhot6/gradle-4.3/lib/gradle-launcher-4.3.jar org.gradle.launcher.daemon.bootstrap.GradleDaemon 4.3
29412 ?        Sl     0:03 /usr/lib/jvm/java-8-oracle/bin/java -Djava.security.manager=worker.org.gradle.process.internal.worker.child.BootstrapSecurityManager -Dshim.classpath=/home/sakurai/.gradle/caches/modules-2/files-2.1/org.clojure/clojure/1.8.0/313066c972b6b00f390674bf0e9f732c07f205a4/clojure-1.8.0.jar:/home/sakurai/.gradle/caches/modules-2/files-2.1/http-kit/http-kit/2.2.0/70b17515f43a0e74937cdc0dea270d4ef13c7f32/http-kit-2.2.0.jar:/home/sakurai/project/olive/clojure/gradle-clojure-samples/build/classes/java/main:/home/sakurai/project/olive/clojure/gradle-clojure-samples/build/resources/main:/home/sakurai/project/olive/clojure/gradle-clojure-samples/src/main/clojure:/home/sakurai/project/olive/clojure/gradle-clojure-samples/build/classes/clojure/main:/home/sakurai/.gradle/caches/modules-2/files-2.1/org.projectodd.shimdandy/shimdandy-impl/1.2.0/46257b0533ce32a80b12aee74210627cea24e08d/shimdandy-impl-1.2.0.jar:/home/sakurai/.gradle/caches/modules-2/files-2.1/io.github.gradle-clojure/gradle-clojure-tools/0.3.0/f2d94bc635e2d809a4164d17f328bd3920dd2da9/gradle-clojure-tools-0.3.0.jar:/home/sakurai/.gradle/caches/modules-2/files-2.1/org.clojure/tools.nrepl/0.2.12/72b4be694c5a9b253e4cd10d512ae60bb80a5450/tools.nrepl-0.2.12.jar:/home/sakurai/.gradle/caches/modules-2/files-2.1/org.clojure/clojure/1.2.0/2b63087f3d0e3076db16d7ede71018d9ea694b89/clojure-1.2.0.jar -Dfile.encoding=UTF-8 -Duser.country=US -Duser.language=en -Duser.variant -cp /home/sakurai/.gradle/caches/4.3/workerMain/gradle-worker.jar worker.org.gradle.process.internal.worker.GradleWorkerMain 'Gradle Worker Daemon 1'
29507 pts/2    S+     0:00 grep --color=auto gradle
sakurai@envyhome ~/project/olive/clojure/gradle-clojure-samples $ ./gradlew assemble

> Task :compileClojure 
Reusing existing Clojure runtime.


BUILD SUCCESSFUL in 0s
3 actionable tasks: 3 executed
sakurai@envyhome ~/project/olive/clojure/gradle-clojure-samples $ unzip -l build/libs/gradle-clojure-samples-0.1.0-SNAPSHOT.jar 
Archive:  build/libs/gradle-clojure-samples-0.1.0-SNAPSHOT.jar
  Length      Date    Time    Name
---------  ---------- -----   ----
        0  2017-11-28 22:34   META-INF/
       25  2017-11-28 22:34   META-INF/MANIFEST.MF
        0  2017-11-28 22:34   sample/
     1142  2017-11-28 22:34   sample/core$bar.class
     2784  2017-11-28 22:34   sample/core__init.class
      974  2017-11-28 22:34   sample/core$foo.class
     2058  2017-11-28 22:34   sample/core$loading__5569__auto____200.class
     1322  2017-11-28 22:34   sample/core$fn__202.class
       11  2017-11-28 22:34   sample.txt
---------                     -------
     8316                     9 files
sakurai@envyhome ~/project/olive/clojure/gradle-clojure-samples $ 
```

**Comment by pbzdyl** [comment]
Thank you @anolivetree, I was able to replicate your issue locally. I think it might be caused by our usage of ClojureWorker/shimdandy that might retain some state used by the compiler (CC @ajoberstar).

Could you please add the following to your `$HOME/.gradle/gradle.properties:

`org.gradle.daemon=false`

Does this configuration change make the issue to disappear? (it did for me) If it does, you can use it as a temporary workaround until we have the fix released.

**Comment by anolivetree** [comment]
Thank you for investigating it. The workaround seems to be working for me too.

**Comment by ajoberstar** [comment]
@anolivetree Can you give 0.3.1-rc.1 a try to see if it resolves this for you?

**Comment by anolivetree** [comment]
@ajoberstar Thank you for the fix. It seems to be working for my Java project.

However, ClojureWorkerClassLoader doesn't work for my Android project with the following error.

```
Caused by: java.lang.RuntimeException: Failed to load ClojureRuntimeShimImpl
        at org.projectodd.shimdandy.ClojureRuntimeShim.newRuntime(ClojureRuntimeShim.java:21)
        at gradle_clojure.plugin.internal.ClojureWorker$ClojureRuntime.get(ClojureWorker.java:94)
        at gradle_clojure.plugin.internal.ClojureWorker.run(ClojureWorker.java:49)
        at org.gradle.workers.internal.DefaultWorkerServer.execute(DefaultWorkerServer.java:36)
        at org.gradle.workers.internal.WorkerDaemonServer.execute(WorkerDaemonServer.java:46)
        at org.gradle.workers.internal.WorkerDaemonServer.execute(WorkerDaemonServer.java:30)
        at org.gradle.process.internal.worker.request.WorkerAction.run(WorkerAction.java:100)
        at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:35)
        at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
        at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:146)
        at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:128)
        at org.gradle.internal.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:404)
        ... 3 more
Caused by: java.lang.SecurityException: Prohibited package name: java.lang
        at gradle_clojure.plugin.internal.ClojureWorkerClassLoader.loadClass(ClojureWorkerClassLoader.java:46)
        at org.projectodd.shimdandy.ClojureRuntimeShim.newRuntime(ClojureRuntimeShim.java:19)
        ... 14 more
```

I'm not sure if it's the correct fix, but following patch resolved the problem for me. Will you add the catch clause if it does no harm?

```
--- a/modules/gradle-clojure-plugin/src/main/java/gradle_clojure/plugin/internal/ClojureWorkerClassLoader.java
+++ b/modules/gradle-clojure-plugin/src/main/java/gradle_clojure/plugin/internal/ClojureWorkerClassLoader.java
@@ -44,7 +44,7 @@ public class ClojureWorkerClassLoader extends URLClassLoader {
     if (clazz == null) {
       try {
         clazz = this.findClass(name);
-      } catch (ClassNotFoundException | NoClassDefFoundError e) {
+      } catch (ClassNotFoundException | NoClassDefFoundError | SecurityException e) {
         if (clazz == null && allowedClass(name)) {
           clazz = this.getParent().loadClass(name);
         }
```

**Comment by ajoberstar** [comment]
Thanks for testing and providing that patch (you're welcome to submit those as PRs in the future)!

It's applied now in #54 and released as 0.3.1-rc.2.

**Comment by anolivetree** [comment]
Thanks. It's working now.
